pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java

```diff
@@ -352,27 +350,77 @@ public synchronized void readEntriesFailed(ManagedLedgerException exception, Obj
 
     }
 
-    private Consumer getNextConsumer() {
+    /**
+     * Broker will first dispatch messages to upper priority-level consumers if they
+     * have permits, else broker will consider next priority level consumers. 
+     * Also on the same priority-level, it selects consumer in round-robin manner. 
+     * </p> 
+     * If subscription has consumer-A with  priorityLevel 1 and Consumer-B with priorityLevel 2 then broker will dispatch 
+     * messages to only consumer-A until it runs out permit and then broker starts dispatching messages to Consumer-B.
+     * 
+     * Consumer PriorityLevel Permits
+     * C1       1             2
+     * C2       1             1
+     * C3       1             1
+     * C4       1             2
+     * C5       1             1
+     * Result of getNextConsumer(): C1, C2, C3, C1, C4, C5, C4
+     * @return
+     */
+    public Consumer getNextConsumer() {
         if (consumerList.isEmpty() || closeFuture != null) {
             // abort read if no consumers are connected or if disconnect is initiated
             return null;
         }
 
-        if (consumerIndex >= consumerList.size()) {
-            consumerIndex = 0;
+        if (currentConsumerRoundRobinIndex >= consumerList.size()) {
+            currentConsumerRoundRobinIndex = 0;
         }
         
-        // find next available unblocked consumer
-        int unblockedConsumerIndex = consumerIndex;
+        // index of resulting consumer which will be returned 
```

Comment:
The logic looks fine here, although it would be helpful to have a comment to explain the algorithm in its entirety. 

Also, it would be good to verify here that there no performance regression in the case when priorities are not used. Eg. running a quick stress test and verify the CPU usage in broker doesn't shoot up compared to before.